### PR TITLE
CloudRendering Platform / python-xlib validation

### PR DESCRIPTION
**Pull request by ekolve**
This PR adds a new module `ai2thor.platform` as well as several classes that correspond to the Unity3D platforms (Linux64, CloudRendering, WebGL, OSXIntel64). Each platfor implements their own `validate` method which is used when finding a suitable build for a user.

The following use cases apply to users on Linux:

1. If a user does not have a Xorg server running, the following exception is thrown:
```
Exception: The following builds were found, but had missing dependencies. Only one valid platform is required to run AI2-THOR.
Platform Linux64 failed validation with the following errors: Invalid display: :0.9. Failed to connect Can't connect to display ":0.9": [Errno 111] Connection refused Linux64 requires a X11 server to be running with GLX. If you have a NVIDIA GPU, please run: sudo ai2thor-xorg start
```

2. If a user passes in an invalid x_display, but they have other valid displays, the following gets presented:
```
Exception: The following builds were found, but had missing dependencies. Only one valid platform is required to run AI2-THOR.
Platform Linux64 failed validation with the following errors: Invalid display, non-existent screen: :0.9
  Linux64 requires a X11 server to be running with GLX. The following valid displays were found :0.0, :0.1
```

3. If a user passes in no x_display, but there are valid displays, then the first one is selected.
4. If CloudRendering is enabled (disabled by default, since its still getting validated) and a CloudRendering build exists it will get selected unless x_display is passed in.

### CODE REVIEW

**Comment by ekolve** [comment]
> This pull request **introduces 1 alert** and **fixes 1** when merging [7dc6c17](https://github.com/allenai/ai2thor/commit/7dc6c17eee4abe9a19bdf596eb88228a78ac7f18) into [6b29a56](https://github.com/allenai/ai2thor/commit/6b29a56866a5a6813f9d1dd09f7efc5c8f45de05) - [view on LGTM.com](https://lgtm.com/projects/g/allenai/ai2thor/rev/pr-fce758d8ae42c0e6b548182cf4b5fbd5c51caad4)
> 
> **new alerts:**
> 
> * 1 for Unused import
> 
> **fixed alerts:**
> 
> * 1 for Unused import

These have been addressed.

File: ai2thor/controller.py [hunk 1]
```diff
@@ -410,6 +409,34 @@ def __init__(
...
+            self.x_display = os.environ["DISPLAY"]
+
+        if self.x_display and ":" not in self.x_display:
+            self.x_display = ":" + self.x_display
+
+        if quality not in QUALITY_SETTINGS:
+            valid_qualities = filter(
+                lambda q: QUALITY_SETTINGS[q] > 0,
+                sorted(QUALITY_SETTINGS.keys(), key=lambda q: QUALITY_SETTINGS[q]),
+            )
```
Comment (Lucaweihs) [inline ai2thor/controller.py:426]: Slightly shorter might be:
```python
valid_qualities = [q for q, v in sorted(QUALITY_SETTINGS.items(), key=lambda qv: qv[1]) if v > 0]
```

File: ai2thor/controller.py [hunk 2]
```diff
@@ -1049,73 +1056,105 @@ def _branch_commits(self, branch):
...
+                    "No build exists for arch=%s and commits: %s"
+                    % (platform.system(), ", ".join(map(lambda x: x[:8], commits)))
+                )
 
-        if commit_build and commit_build.commit_id != commits[0]:
-            warnings.warn(
-                "Build for the most recent commit: %s is not available.  Using commit build %s"
-                % (commits[0], commit_build.commit_id)
+        # select the first build + platform that succeeds
+        # For Linux, this will select CloudRendering, then Linux64 (X11)
```
Comment (Lucaweihs) [inline ai2thor/controller.py:1112]: Currently we don't select cloud rendering first right? As it's still being validated?

File: ai2thor/platform.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
+import Xlib.display
+import glob
+import warnings
+import os
+import ctypes.util
+import xml.etree.ElementTree
+
+
+class Request:
```
Comment (Lucaweihs) [inline ai2thor/platform.py:9]: Can we give this a slightly more descriptive name? Perhaps `PlatformRequest` or `BuildRequest`?

**Review by Lucaweihs — APPROVED** [review]
Looks fantastic, I really appreciate all of the super informative error messages. I added a few very minor comments. Do we have any documentation anywhere that describes how THOR falls back on various displays/builds depending on what's available? I could see this being useful as there seems to be a growing list of options that differs by platform.

File: ai2thor/platform.py (hunk 3 above)
Comment (ekolve) [reply ai2thor/platform.py:9]: My argument for using just the name `Request` as opposed to `PlatformRequest` is that the class resides within the `ai2thor.platform` module.  When we use this class, we use it as `ai2thor.platform.Request` (within controller.py) so there shouldn't be any ambiguity about meaning.  As well, this is consistent with how the Python standard library names classes. Consider [Request](https://docs.python.org/3/library/urllib.request.html#request-objects).

File: ai2thor/controller.py (hunk 2 above)
Comment (ekolve) [reply ai2thor/controller.py]: I updated the comment to reflect the new behavior. CloudRendering is disabled by default.

File: ai2thor/controller.py (hunk 1 above)
Comment (ekolve) [reply ai2thor/controller.py]: Nice! I updated the line.

**Pull request merged**